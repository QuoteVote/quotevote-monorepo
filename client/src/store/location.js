/**
 * Location Redux Store
 * Manages user's current location and permission state
 */

import { createSlice } from '@reduxjs/toolkit';
import { getCurrentPosition, checkPermission, GeoLocationError } from '../utils/geolocationService';

const initialState = {
  currentLocation: null, // { lat, lng }
  permissionState: 'prompt', // 'prompt' | 'granted' | 'denied' | 'unsupported'
  isLoading: false,
  error: null, // { type, message }
  lastUpdated: null,
};

const locationSlice = createSlice({
  name: 'location',
  initialState,
  reducers: {
    setLocation: (state, action) => {
      state.currentLocation = action.payload;
      state.lastUpdated = Date.now();
      state.error = null;
    },
    setPermissionState: (state, action) => {
      state.permissionState = action.payload;
    },
    setLoading: (state, action) => {
      state.isLoading = action.payload;
    },
    setError: (state, action) => {
      state.error = action.payload;
      state.isLoading = false;
    },
    clearError: (state) => {
      state.error = null;
    },
    clearLocation: (state) => {
      state.currentLocation = null;
      state.lastUpdated = null;
      state.error = null;
    },
    reset: () => initialState,
  },
});

export const {
  setLocation,
  setPermissionState,
  setLoading,
  setError,
  clearError,
  clearLocation,
  reset,
} = locationSlice.actions;

// Thunks

/**
 * Fetch user's current location
 * @param {object} options - Options passed to getCurrentPosition
 */
export const fetchLocation = (options = {}) => async (dispatch) => {
  dispatch(setLoading(true));
  dispatch(clearError());

  try {
    const location = await getCurrentPosition(options);
    dispatch(setLocation(location));
    dispatch(setPermissionState('granted'));
    return location;
  } catch (error) {
    dispatch(setError({
      type: error.type || GeoLocationError.UNKNOWN,
      message: error.message || 'Failed to get location',
    }));
    
    if (error.type === GeoLocationError.PERMISSION_DENIED) {
      dispatch(setPermissionState('denied'));
    }
    
    throw error;
  } finally {
    dispatch(setLoading(false));
  }
};

/**
 * Check and update permission state
 */
export const updatePermissionState = () => async (dispatch) => {
  try {
    const state = await checkPermission();
    dispatch(setPermissionState(state));
    return state;
  } catch (error) {
    console.error('Failed to check permission:', error);
    return 'prompt';
  }
};

// Selectors

export const selectLocation = (state) => state.location.currentLocation;
export const selectPermissionState = (state) => state.location.permissionState;
export const selectIsLoading = (state) => state.location.isLoading;
export const selectError = (state) => state.location.error;
export const selectLastUpdated = (state) => state.location.lastUpdated;

export const selectHasLocation = (state) => state.location.currentLocation !== null;
export const selectIsPermissionGranted = (state) => state.location.permissionState === 'granted';
export const selectIsPermissionDenied = (state) => state.location.permissionState === 'denied';

export default locationSlice.reducer;
