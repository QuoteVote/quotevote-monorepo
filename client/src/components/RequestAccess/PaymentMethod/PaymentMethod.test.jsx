import React from 'react'
import { render, act } from '@testing-library/react'
import { Provider } from 'react-redux'
import PaymentMethod from './PaymentMethod'
import store from '../../../store/store'

const props = {
  cardDetails: { brand: 'visa', last4: '4242' },
  setCardDetails: () => {},
  isContinued: false,
  onSubmit: () => {},
}

const PaymentMethodWrapper = () => (
  <Provider store={store}>
    <PaymentMethod {...props} />
  </Provider>
)

describe('PaymentMethod test -', () => {
  it('renders correctly', async () => {
    let container
    await act(async () => {
      const res = render(<PaymentMethodWrapper />)
      container = res.container
    })
    expect(container.firstChild).toMatchSnapshot()
  })
})
