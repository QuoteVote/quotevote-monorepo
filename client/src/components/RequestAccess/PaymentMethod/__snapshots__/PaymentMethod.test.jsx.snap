// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`PaymentMethod test - > renders correctly 1`] = `
<form>
  <div
    class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-o5b05n-MuiPaper-root-MuiCard-root"
  >
    <div
      class="MuiCardHeader-root css-149tc0y-MuiCardHeader-root"
    >
      <div
        class="MuiCardHeader-avatar css-1r9wl67-MuiCardHeader-avatar"
      >
        <p
          class="MuiTypography-root MuiTypography-body1 makeStyles-stepNumber-4 css-1edfpdg-MuiTypography-root"
        >
          2
        </p>
      </div>
      <div
        class="MuiCardHeader-content css-1jkp9bu-MuiCardHeader-content"
      >
        <p
          class="MuiTypography-root MuiTypography-body1 makeStyles-stepName-5 css-1edfpdg-MuiTypography-root"
        >
          Would you like to support this app?
        </p>
      </div>
    </div>
    <input
      name="cardNumber"
      type="hidden"
    />
    <input
      name="expiry"
      type="hidden"
    />
    <input
      name="cvv"
      type="hidden"
    />
  </div>
</form>
`;
